File: python-single/src/klv/common/pathId.py

from klv import Walker
from klv.common.enums.pathType import PathType


class PathId:
  NONE =  "U/65535/0"
  CURRENT = "U/65535/1"

  def __init__(self, walker: Walker):
    self.raw = walker.read_u32()

  def pathType(self):
    return PathType(self.raw >> 24) if self.raw >> 24 > 0 else PathType.U

  def number(self):
    return self.raw >> 8 & 0x0000ffff if self.raw >> 24 > 0 else 0xffff

  def component(self):
    return self.raw & 0x000000ff if self.raw >> 24 > 0 else 0

  def getHasPath(self):
    return self.to_string() != ""

  def getIsChannel(self):
    return self.pathType() is PathType.CH

  def getIsAux(self):
    return self.pathType() is PathType.AUX

  def getIsMain(self):
    return self.pathType() is PathType.MN

  def getIsGroup(self):
    return self.pathType() is PathType.GP

  def getIsMixMinus(self):
    return self.pathType() is PathType.MM

  def getIsTrack(self):
    return self.pathType() is PathType.TK

  def getIsDirectOutput(self):
    return self.pathType() is PathType.DIR

  def getIsVCAMaster(self):
    return self.pathType() is PathType.VCA

  def getIsRemoteFader(self):
    return self.pathType() is PathType.REM_FADER

  def getIsRemoteAux(self):
    return self.pathType() is PathType.REM_AUX

  def getIsRemote(self):
    return self.getIsRemoteFader() or self.getIsRemoteAux()

  def to_string(self):
    path = "{}/{}/{}".format(self.pathType(), self.number(), self.component())

    return "" if path == self.NONE or path == self.CURRENT else path

File: java-single/app/src/main/java/javaSingle/klv/common/PathId.java
package javaSingle.klv.common;

import io.netty.buffer.ByteBuf;
import javaSingle.klv.common.enums.PanType;
import javaSingle.klv.common.enums.PathType;

import java.util.EnumSet;

public class PathId {
  public static final int RAW_NO_PATH = 0xffff00;
  public static final String NONE = "U/65535/0";
  public static final String CURRENT = "U/65535/1";

  public long raw;

  public PathId(ByteBuf buf) {
    this.raw = buf.readUnsignedIntLE();
  }

  public PathId(long raw) {
    this.raw = (raw >> 24 > 0 ? raw : RAW_NO_PATH);
  }

  public PathId(PathType pathType, int number, int component) {
    raw = (pathType.ordinal() << 24) + (number << 8) + component;
  }

  public PathId(String s) {
    String[] token = (s.isEmpty() ? NONE : s).split("/");

    raw =
      (PathType.valueOf(token[0]).ordinal() << 24) +
      (Integer.parseInt(token[1]) << 8) +
      (token.length == 2 ? 0 : Integer.parseInt(token[2]));
  }

  public boolean isChannel() {
    return raw >> 24 == 1;
  }

  public boolean isValid() {
    return raw != RAW_NO_PATH;
  }

  @Override
  public boolean equals(Object o) {
    return (o != null && (o == this || o.getClass() == getClass() && o.hashCode() == hashCode()));
  }

  @Override
  public int hashCode() {
    return (int) raw;
  }

  public PathType type() {
    return (raw >> 24 > 0 ? PathType.values()[(int) raw >> 24] : PathType.U);
  }

  public int number() {
    return (raw >> 24 > 0 ? (int) raw >> 8 & 0x0000ffff : 0xffff);
  }

  public int component() {
    return (raw >> 24 > 0 ? (int) raw & 0x000000ff : 0);
  }

  public PathId component(int component) {
    return new PathId(type(), number(), component);
  }

  @Override
  public String toString() {
    String path = String.format("%s/%s/%s", type(), number(), component());

    return ((path.equals(NONE) || path.equals(CURRENT)) ? "" : path);
  }

  public boolean isBuss() {
    return EnumSet.of(PathType.MN, PathType.GP, PathType.TK, PathType.AUX, PathType.MM).contains(type());
  }


  public String getPath() {
    return toString();
  }

  public Boolean getHasPath() {
    return toString() != "";
  }

  public Boolean getIsChannel() {
    return isChannel();
  }

  public Boolean getIsAux() {
    return type() == PathType.AUX;
  }

  public Boolean getIsMain() {
    return type() == PathType.MN;
  }

  public Boolean getIsGroup() {
    return type() == PathType.GP;
  }

  public Boolean getIsMixMinus() {
    return type() == PathType.MM;
  }

  public Boolean getIsTrack() {
    return type() == PathType.TK;
  }

  public Boolean getIsDirectOutput() {
    return type() == PathType.DIR;
  }

  public Boolean getIsVCAMaster() {
    return type() == PathType.VCA;
  }

  public Boolean getIsRemoteFader() {
    return type() == PathType.REM_FADER;
  }

  public Boolean getIsRemoteAux() {
    return type() == PathType.REM_AUX;
  }

  public Boolean getIsRemote() {
    return getIsRemoteFader() || getIsRemoteAux();
  }

  public Boolean isSpill() {
    return component() > 0;
  }

  public boolean isCurrentPath() {
    return type().equals(PathType.U) && ((int) raw & 0x000000ff) == 1;
  }

  public String getName() {
    String type = "";

    if (getIsMain()) type = "Main";
    if (getIsGroup()) type = "Group";
    if (getIsAux()) type = "Aux";
    if (getIsTrack()) type = "Track";
    if (getIsMixMinus()) type = "Mix Minus";

    return type + " " + (number() + 1);
  }

  public String getShortName() {
    String type = "";

    if (getIsMain()) type = "M";
    if (getIsGroup()) type = "G";
    if (getIsAux()) type = "A";
    if (getIsTrack()) type = "T";
    if (getIsMixMinus()) type = "MM";

    return type + (number() + 1);
  }

  public Integer getPanIndex() {
    Integer panIndex = PanType.PATH_PANNER.getPanIndex();
    if (getIsTrack()) panIndex = PanType.TRACK_PANNER.getPanIndex();
    if (getIsAux()) panIndex = PanType.AUX_PANNER.getPanIndex();
    return panIndex;
  }
}
